h/schemas/forms/accounts/login.py

```diff
@@ -9,14 +9,27 @@
 _ = i18n.TranslationString
 
 
+@colander.deferred
+def _deferred_username_widget(node, kw):
+    """Return a username widget that autofocuses if username isn't pre-filled."""
+    return deform.widget.TextInputWidget(autofocus=_should_autofocus_username(kw))
+
+
+@colander.deferred
+def _deferred_password_widget(node, kw):
+    """Return a password widget that autofocuses if username *is* pre-filled."""
+    return deform.widget.PasswordWidget(autofocus=not _should_autofocus_username(kw))
+
+
 class LoginSchema(CSRFSchema):
+
     username = colander.SchemaNode(
         colander.String(),
         title=_("Username / email"),
-        widget=deform.widget.TextInputWidget(autofocus=True),
+        widget=_deferred_username_widget,
```

Comment:
"Deferred" means that Colander will call the `_deferred_username_widget()` function to get the username widget _when the schema is bound to a request_ rather than when the `LoginSchema` object is first initialized. It means we can vary what username widget we return based on the request.

Reply:
Ok, so we define our forms in Python? Did not see that one coming. I was fully expecting this to be front-end.